server/ocsp.go

```diff
@@ -400,15 +409,248 @@ func (srv *Server) NewOCSPMonitor(tc *tls.Config) (*tls.Config, *OCSPMonitor, er
+				// GetClientCertificate is used by a client to send the client cert
+				// to a server. We're a server, so we must not set this.
+				tc.GetClientCertificate = nil
+				sopts.LeafNode.TLSConfig = tc
+			},
+		}
+		configs = append(configs, o)
+	}
+	for i, remote := range sopts.LeafNode.Remotes {
+		opts := remote.tlsConfigOpts
+		if config := remote.TLSConfig; config != nil {
+			o := &tlsConfigKind{
+				kind:      typeStringMap[LEAF],
+				tlsConfig: config,
+				tlsOpts:   opts,
+				apply: func(tc *tls.Config) {
+					// GetCertificate is used by a server to send the server cert to a
+					// client. We're a client, so we must not set this.
+					tc.GetCertificate = nil
+
+					sopts.LeafNode.Remotes[i].TLSConfig = tc
+				},
+			}
+			configs = append(configs, o)
+		}
+	}
+	if config := sopts.Gateway.TLSConfig; config != nil {
+		opts := sopts.Gateway.tlsConfigOpts
+		o := &tlsConfigKind{
+			kind:      typeStringMap[GATEWAY],
+			tlsConfig: config,
+			tlsOpts:   opts,
+			apply:     func(tc *tls.Config) { sopts.Gateway.TLSConfig = tc },
+		}
+		configs = append(configs, o)
+	}
+	for i, remote := range sopts.Gateway.Gateways {
+		opts := remote.tlsConfigOpts
+		if config := remote.TLSConfig; config != nil {
+			o := &tlsConfigKind{
+				kind:      typeStringMap[GATEWAY],
+				tlsConfig: config,
+				tlsOpts:   opts,
+				apply: func(tc *tls.Config) {
+					sopts.Gateway.Gateways[i].TLSConfig = tc
+				},
+			}
+			configs = append(configs, o)
+		}
+	}
+	return configs
+}
+
+func (s *Server) enableOCSP() error {
+	configs := s.configureOCSP()
+
+	for _, config := range configs {
+		tc, mon, err := s.NewOCSPMonitor(config)
+		if err != nil {
+			return err
+		}
+		// Check if an OCSP stapling monitor is required for this certificate.
+		if mon != nil {
+			s.ocsps = append(s.ocsps, mon)
+
+			// Override the TLS config with one that follows OCSP.
+			config.apply(tc)
+		}
+	}
+
+	return nil
+}
+
+func (s *Server) startOCSPMonitoring() {
+	s.mu.Lock()
+	ocsps := s.ocsps
+	s.mu.Unlock()
+	if ocsps == nil {
+		return
+	}
+	for _, mon := range ocsps {
+		m := mon
+		m.mu.Lock()
+		kind := m.kind
+		m.mu.Unlock()
+		s.Noticef("OCSP Stapling enabled for %s connections", kind)
+		s.startGoRoutine(func() { m.run() })
+	}
+}
+
+func (s *Server) reloadOCSP() error {
+	if err := s.setupOCSPStapleStoreDir(); err != nil {
+		return err
+	}
+
+	s.mu.Lock()
+	ocsps := s.ocsps
+	s.mu.Unlock()
+
+	// Stop all OCSP Stapling monitors in case there were any running.
+	for _, oc := range ocsps {
+		oc.stop()
+	}
+
+	configs := s.configureOCSP()
+
+	// Restart the monitors under the new configuration.
+	ocspm := make([]*OCSPMonitor, 0)
+	for _, config := range configs {
+		tc, mon, err := s.NewOCSPMonitor(config)
+		if err != nil {
+			return err
+		}
+		// Check if an OCSP stapling monitor is required for this certificate.
+		if mon != nil {
+			ocspm = append(ocspm, mon)
+
+			// Apply latest TLS configuration.
+			config.apply(tc)
+		}
+	}
```

Comment:
In case the cert does not have the must staple flag and running in auto mode which is the default, then it would get disabled here: https://github.com/nats-io/nats-server/pull/2263/files#diff-7bc3c6340f0feb2d8426342029fd9c192a397b1ade67c05e25a90d076c39248bR359